enhancements/cloud-integration/infrastructure-external-platform-type.md

```diff
@@ -0,0 +1,513 @@
+- Should we gate the "External" platform addition behind the feature gate by generating separate CRD for TPNU clusters?
+  - Do we already have all the necessary mechanisms for gating such API changes behind the feature gate?
+- Should we invest in preparing workflows that will perform UPI with the "None"/"External" platform types installation
+  on the AWS or GCP, or existed vSphere-based workflows would be enough?
+
+### Test Plan
+
+During the first stages, we must ensure that OpenShift [built-in operators whose behaviour depends on platform type](#affected-components) treat the "External" platform type the same way as "None".
+To achieve this - existing infrastructure and mechanisms employed for exercising topologies with the "None" platform type might be used.
+
+For the moment of writing, the only workflow where the "None" platform tests is [upi-conf-vsphere-platform-none-ovn](https://steps.ci.openshift.org/workflow/upi-vsphere-platform-none-ovn).
+Based on this workflow, a new one with platform type "External" with a respective set of jobs should be added to ensure that we do not disrupt the current openshift operation.
+
+However, given that vSphere is the only platform where we're exercising clusters installation with platform "None" specified,
+it would be beneficial to develop additional workflows using a provider with better capacity and performance (AWS or GCP, perhaps).
+
+### Graduation Criteria
+
+Given that `infrastructure.config.openshift.io` has already been released and has to be supported during the whole major release lifecycle,
+this change will be GAed from the beginning.
+
+However, components behaviour changes might be gated with the feature gates.
+Specific graduation criteria for each component should be defined separately in respective enhancement documents.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+This enhancement does not introduce any additional components, it just describes changes in "openshift/api".
+
+PlatformType is expected to be set once during cluster installation and does not expect to be changed, so adding new platform type
+should not affect upgrade/downgrade process across existing clusters.
+
+### Version Skew Strategy
+
+Since setting PlatformType is a day 0 operation, sets up during cluster installation and does not expect to be changed during cluster lifecycle,
+version skew should not be the case there.
+
+### Operational Aspects of API Extensions
+
+It should be no different from other PlatformTypes. Scalability, API throughput and availability should not be affected.
+In the first phases, it is expected to work the same way as for PlatformType "None"
+
+#### Failure Modes
+
+- OpenShift built-in operators would not recognize the new PlatformType and would go into Degraded or crashed.
+  This will break new clusters installations with that new "External" platform type.
```

Comment:
Could you go through the operators first and check how they handle an unknown value?

Reply:
this will be part of the phase 2 work, essentially discovering how they react and ensuring it's not detrimental.

Reply:
What if we add the new ExternalPlatformSpec and ExternalPlatformStatus fields, but we use those fields optionally when the platform type is `None` and we don't add the new enum value?

Reply:
I would recommend this be pre-requisite work for phase 1.

If an operator today has a switch case on the platform, that has a specific `None` case, we need to add the new value to that. If the `default` case is to error, that likely means that the cluster will go degraded right? Does it matter, it's new External stuff so it won't break anyones upgrades, but it couldn't go GA like that.

I do wonder about whether this new Enum value should be TP in the first instance. What's the advantage of GAing it at phase 0 as opposed to phase 1?
TP would allow you to defer some of this until you declare it GA

Reply:
i'm a little confused, @dhellmann are you suggesting we should not add the `External` platform type at all?

Reply:
we talked with @JoelSpeed a little today about his comment. just to clarify, Joel is asking for us to add more information about how the operators would respond today if they get an unrecognized platform. @lobziik has added this information to most of the specific component descriptions, although i know he is going to review the CCM parts.

Reply:
@elmiko, I was reading the proposal through the lens of doing the least work possible. It seems like we only have to touch some operators to add a new item to a case statement, and we could avoid that if we didn't add a new enum value. Keeping the rest of the API changes would let us update the operators that need to behave differently when there is an external cloud provider.

Reply:
Would doing so deviate from the pattern of the discriminated union that is already in place within the ProviderSpec and ProviderStatus?

Does it make the API implicit? This platform is external if and only if you have external config under the `none` platform, I would probably lean toward keeping the value and making sure it's declarative that this platform is external

Reply:
>What if we add the new ExternalPlatformSpec and ExternalPlatformStatus fields, but we use those fields optionally when the platform type is None and we don't add the new enum value?

+1 to what @JoelSpeed said. This will add `None` platform additional meanings and behaviour. 
Another thing which i'm concerned about is existing clusters. We should be more careful during upgrades and with version skew handling due to there would be some number of existing clusters. With new value in the enum, we can more-less confidently say that it would be only fresh installations.

Reply:
> Does it make the API implicit? This platform is external if and only if you have external config under the `none` platform, I would probably lean toward keeping the value and making sure it's declarative that this platform is external

A bit? I was looking at it as redefining "None" to mean "not one of ours" and then in that external section we might have a flag to indicate whether there's an external provider or not.

It sounds like the group doesn't think it's a good idea, though, so that's fine.

Reply:
@lobziik and i talked about this a little and i tend to agree with his reasoning here, i prefer to have both "none" and "external". i do appreciate what you are asking though @dhellmann, it gave us good caution for thought.